src/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -282,262 +280,530 @@ internal static ulong ToUInt64(object value)
+            return false;
+        }
 
-            char firstNonWhitespaceChar = value[firstNonWhitespaceIndex];
-            if (char.IsDigit(firstNonWhitespaceChar) || firstNonWhitespaceChar == '-' || firstNonWhitespaceChar == '+')
+        private static bool TryParseRareEnum<TResult>(RuntimeType enumType, string originalValueString, ReadOnlySpan<char> value, bool ignoreCase, bool throwOnFailure, out TResult enumResult)
+        {
+            if (StartsNumber(value[0]))
             {
                 Type underlyingType = GetUnderlyingType(enumType);
-                object temp;
-
                 try
                 {
-                    value = value.Trim();
-                    temp = Convert.ChangeType(value, underlyingType, CultureInfo.InvariantCulture);
-                    parseResult.parsedEnum = ToObject(enumType, temp);
+                    enumResult = (TResult)ToObject(enumType, Convert.ChangeType(value.ToString(), underlyingType, CultureInfo.InvariantCulture));
                     return true;
                 }
                 catch (FormatException)
-                { // We need to Parse this as a String instead. There are cases
-                  // when you tlbimp enums that can have values of the form "3D".
-                  // Don't fix this code.
+                {
+                    // We need to Parse this as a String instead. There are cases
+                    // when you tlbimp enums that can have values of the form "3D".
                 }
-                catch (Exception ex)
+                catch when (!throwOnFailure)
                 {
-                    if (parseResult.canThrow)
-                        throw;
-                    else
-                    {
-                        parseResult.SetFailure(ex);
-                        return false;
-                    }
+                    enumResult = default;
+                    return false;
                 }
             }
 
-            // Find the field. Let's assume that these are always static classes 
-            // because the class is an enum.
-            TypeValuesAndNames entry = GetCachedValuesAndNames(rtType, true);
+            if (TryParseByName(enumType, originalValueString, value, ignoreCase, throwOnFailure, out ulong ulongResult))
+            {
+                try
+                {
+                    enumResult = (TResult)ToObject(enumType, ulongResult);
+                    return true;
+                }
+                catch when (!throwOnFailure) { }
+            }
+
+            enumResult = default;
+            return false;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static bool StartsNumber(char c) => char.IsInRange(c, '0', '9') || c == '-' || c == '+';
+
+        private static bool TryParseByName(RuntimeType enumType, string originalValueString, ReadOnlySpan<char> value, bool ignoreCase, bool throwOnFailure, out ulong ulongResult)
```

Comment:
Do we really need the parsed flag? How about something like this, where we have a `ParsedFailed` helper?

```C#
        private static bool TryParseByName(RuntimeType enumType, string originalValueString, ReadOnlySpan<char> value, bool ignoreCase, bool throwOnFailure, out ulong ulongResult)
        {
            // Find the field. Let's assume that these are always static classes because the class is an enum.
            TypeValuesAndNames entry = GetCachedValuesAndNames(enumType, getNames: true);
            string[] enumNames = entry.Names;
            ulong[] enumValues = entry.Values;

            ulong result = 0;
            while (value.Length > 0)
            {
                // Find the next separator.
                ReadOnlySpan<char> subvalue;
                int endIndex = value.IndexOf(enumSeparatorChar);
                if (endIndex == -1)
                {
                    // No next separator; use the remainder as the next value.
                    subvalue = value.Trim();
                    value = default;
                }
                else if (endIndex != value.Length - 1)
                {
                    // Found a separator before the last char.
                    subvalue = value.Slice(0, endIndex).Trim();
                    value = value.Slice(endIndex + 1);
                }
                else
                {
                    // Last char was a separator, which is invalid.
                    return ParseFailed(throwOnFailure, originalValueString, out ulongResult);
                }

                // Try to match this substring against each enum name
                bool success = false;
                if (ignoreCase)
                {
                    for (int i = 0; i < enumNames.Length; i++)
                    {
                        if (subvalue.EqualsOrdinalIgnoreCase(enumNames[i]))
                        {
                            result |= enumValues[i];
                            success = true;
                            break;
                        }
                    }
                }
                else
                {
                    for (int i = 0; i < enumNames.Length; i++)
                    {
                        if (subvalue.EqualsOrdinal(enumNames[i]))
                        {
                            result |= enumValues[i];
                            success = true;
                            break;
                        }
                    }
                }

                if (!success)
                {
                    return ParseFailed(throwOnFailure, originalValueString, out ulongResult);
                }
            }

           ulongResult = result;
           return true;
        }
        
        // Or this is a `goto Failed` at the end of the method above.
        bool ParseFailed(bool throwOnFailure, string originalValueString, out ulong ulongResult)
        {
            if (throwOnFailure)
            {
                throw new ArgumentException(SR.Format(SR.Arg_EnumValueNotFound, originalValueString));
            }

            ulongResult = default;
            return false;
        }
```

Reply:
> bool ParseFailed(bool throwOnFailure, string originalValueString, out ulong ulongResult)

You can't use `out` in a local function.  And I don't want to use captures and force these to be hoisted to a struct.

> Or this is a `goto Failed` at the end of the method above.

Would `goto` here make a measurably beneficial impact on throughput?  If not, I'd prefer to avoid it.

Reply:
> Would `goto` here make a measurably beneficial impact on throughput?  If not, I'd prefer to avoid it.

It seems like we just save a single if check (outside the loop) so probably not worth it. I just prefer not seeing multiple flags used for control flow. I also **initially** saw the `success` flag as unnecessary, but couldn't figure out a way to remove it (I was thinking something like initializing result to -1, but its a ulong, and something like ulong.MaxValue breaks the or'ing logic, which requires the start to be at 0).

Maybe something like this?
```C#
// Try to match this substring against each enum name
int i = 0;
if (ignoreCase)
{
    for (; i < enumNames.Length; i++)
    {
        if (subvalue.EqualsOrdinalIgnoreCase(enumNames[i]))
        {
            result |= enumValues[i];
            break;
        }
    }
}
else
{
    for (; i < enumNames.Length; i++)
    {
        if (subvalue.EqualsOrdinal(enumNames[i]))
        {
            result |= enumValues[i];
            break;
        }
    }
}

// Not sure if this condition is correct
if (i == 0 || i == enumNames.Length)
{
    return ParseFailed(throwOnFailure, originalValueString, out ulongResult);
}
```

> You can't use `out` in a local function.  And I don't want to use captures and force these to be hoisted to a struct.

It would work if we made the local a static method right? Can you explain the concern with doing that?

Reply:
> It would work if we made the local a static method right? Can you explain the concern with doing that?

That moves private implementation details of this method out into a larger scope where it's not relevant.  If it had significant benefits that would outweigh the downsides, then sure, but I'm not seeing it.  Not everyone has your same love of goto :wink: